k9mail-library/src/test/java/com/fsck/k9/mail/filter/LineWrapOutputStreamTest.java

```diff
@@ -0,0 +1,308 @@
+        fillWithStaticData(writer);
+        checkWrapsTight(MAX_BYTES_PER_LINE, baos);
+    }
+
+    private static void fillWithStaticData(BufferedWriter writer) throws IOException {
+        for (String field : arrayOfStrings) {
+            writer.write(field);
+        }
+        writer.flush();
+    }
+
+    private static void testTightWrapRandom(final int MAX_BYTES_PER_LINE
+            , ByteArrayOutputStream baos, Random rand) throws IOException {
+        baos.reset();
+        LineWrapOutputStream lwops = new LineWrapOutputStream(baos, MAX_BYTES_PER_LINE);
+        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(lwops), 100);
+        fillWithRandomData(writer, MAX_BYTES_PER_LINE, rand);
+        checkWrapsTight(MAX_BYTES_PER_LINE, baos);
+    }
+
+    private static void fillWithRandomData(BufferedWriter writer, int max_bytes_per_line
+            , Random rand) throws IOException {
+        final int NUM_LINES = 100 + rand.nextInt(400);
+        for (int i = 0; i < NUM_LINES; ++i) {
+            writeRandomLine(writer, max_bytes_per_line, rand);
+        }
+        writer.flush();
+    }
+
+    private static void writeRandomLine(BufferedWriter writer, int MAX_BYTES_PER_LINE
+            , Random rand) throws IOException {
+        final int MAX_WORD_LEN =  2 + rand.nextInt(MAX_BYTES_PER_LINE - 5);
+        int nextWordPos = 2 + rand.nextInt(MAX_WORD_LEN);
+        for (int i = 0; i < MAX_BYTES_PER_LINE; ++i) {
+            if (nextWordPos <= i) {
+                writer.write(' ');
+                nextWordPos += 2 + rand.nextInt(MAX_WORD_LEN);
+            } else {
+                int nextByte = rand.nextInt(64000);
+                while (nextByte == '\n' || nextByte == '\r') nextByte = rand.nextInt(64000);
+                writer.write(nextByte);
+            }
+        }
+    }
+
+    /**
+     * Random bytes because it is hard to count the bytes in preset strings
+     * which are the datums on which wrapping occurs.
+     * CRLF characters are filtered from the input since we will be
+     * assessing optimal, non-overflowing, wrapping and empty, excessive, breaks
+     * would interfere.
+     * @throws Exception
+     */
+    @Test
+    public void testTightWrappingWithRandBytes() throws Exception {
+        for (int i = 0; i < 10; ++i) {
+            testTightWrapRandom(20, baos, rand);
+            testTightWrapRandom(40, baos, rand);
+            testTightWrapRandom(60, baos, rand);
+            testTightWrapRandom(80, baos, rand);
+            testTightWrapRandom(100, baos, rand);
+            testTightWrapRandom(120, baos, rand);
+            testTightWrapRandom(240, baos, rand);
+            testTightWrapRandom(480, baos, rand);
+            testTightWrapRandom(1024, baos, rand);
+            testTightWrapRandom(2048, baos, rand);
+        }
+    }
+
+    private static void checkWrapsTight(int MAX_BYTES_PER_LINE, ByteArrayOutputStream baos) {
+        byte bArray[] = baos.toByteArray();
```

Comment:
I could argue, or I could remove it, it's certainly more independent than the convoluted stream filter it's testing, I can't see the utility of the filter at the byte level unless it is to get packets undersize for a network which I assume it must be with line length 1000, I need to study the protocol. Surely my code isn't that bad?